File: Problems/Algorithms/713. Subarray Product Less Than K/SubarrayProduct.kt
class Solution {
    fun numSubarrayProductLessThanK(nums: IntArray, k: Int): Int {
        if (k == 0) return 0
        
        val n = nums.size
        
        var prod = 1
        var j = 0
        var ans = 0
        for (i in 0..n-1) {
            prod *= nums[i]
            if (prod < k) {
                ans += i - j + 1
            } else {
                while (prod >= k && j <= i) {
                    prod /= nums[j]
                    j++
                }
                ans += i - j + 1
            }
        }
        
        return ans
    }
}

File: Problems/Algorithms/713. Subarray Product Less Than K/subarray_product.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0

        ans = 0
        
        product = 1
        i = 0
        for j, num in enumerate(nums):
            product *= num
            while product >= k:
                product //= nums[i]
                i += 1
            ans += j - i + 1

        return ans
